rl_algorithms/recurrent/learner.py

```diff
@@ -0,0 +1,174 @@
+from typing import Tuple, Union
+
+import numpy as np
+import torch
+import torch.nn as nn
+from torch.nn.utils import clip_grad_norm_
+import torch.optim as optim
+
+from rl_algorithms.common.abstract.learner import Learner, TensorTuple
+import rl_algorithms.common.helper_functions as common_utils
+from rl_algorithms.common.networks.brain import GRUBrain
+from rl_algorithms.dqn.learner import DQNLearner
+from rl_algorithms.registry import LEARNERS, build_loss
+from rl_algorithms.utils.config import ConfigDict
+
+
+@LEARNERS.register_module
+class R2D1Learner(DQNLearner):
+    """Learner for R2D1 Agent.
+
+    Attributes:
+        args (argparse.Namespace): arguments including hyperparameters and training settings
+        hyper_params (ConfigDict): hyper-parameters
+        log_cfg (ConfigDict): configuration for saving log and checkpoint
+        dqn (nn.Module): dqn model to predict state Q values
+        dqn_target (nn.Module): target dqn model to predict state Q values
+        dqn_optim (Optimizer): optimizer for training dqn
+
+    """
+
+    def __init__(
+        self,
+        args: argparse.Namespace,
+        env_info: ConfigDict,
+        hyper_params: ConfigDict,
+        log_cfg: ConfigDict,
+        backbone: ConfigDict,
+        head: ConfigDict,
+        optim_cfg: ConfigDict,
+        device: torch.device,
+    ):
+        DQNLearner.__init__(
+            self,
+            args,
+            env_info,
+            hyper_params,
+            log_cfg,
+            backbone,
+            head,
+            optim_cfg,
+            device,
+        )
+
+        self.backbone_cfg = backbone
+        self.head_cfg = head
+        self.head_cfg.configs.state_size = self.env_info.observation_space.shape
+        self.head_cfg.configs.output_size = self.env_info.action_space.n
+        self.optim_cfg = optim_cfg
+        self.use_n_step = self.hyper_params.n_step > 1
+
+        self._init_network()
+
+    # pylint: disable=attribute-defined-outside-init
+    def _init_network(self):
+        """Initialize networks and optimizers."""
+        self.dqn = GRUBrain(self.backbone_cfg, self.head_cfg).to(self.device)
+        self.dqn_target = GRUBrain(self.backbone_cfg, self.head_cfg).to(self.device)
+        self.loss_fn = build_loss(self.hyper_params.loss_type)
+
+        self.dqn_target.load_state_dict(self.dqn.state_dict())
+
+        # create optimizer
+        self.dqn_optim = optim.Adam(
+            self.dqn.parameters(),
+            lr=self.optim_cfg.lr_dqn,
+            weight_decay=self.optim_cfg.weight_decay,
+            eps=self.optim_cfg.adam_eps,
+        )
+
+        # load the optimizer and model parameters
+        if self.args.load_from is not None:
+            self.load_params(self.args.load_from)
+
+    def update_model(
+        self, experience: Union[TensorTuple, Tuple[TensorTuple]]
+    ) -> Tuple[torch.Tensor, torch.Tensor, list, np.ndarray]:  # type: ignore
+        """Update dqn and dqn target"""
```

Comment:
should modify description.
should remove an empty line.

Reply:
resolved in commit 0949a63. It didn't seems to be fixed.